modules/plugin/geopkg/src/main/java/org/geotools/geopkg/GeoPackage.java

```diff
@@ -1435,11 +1435,13 @@ static TileEntry createTileEntry(ResultSet rs, Connection cx) throws SQLExceptio
         TileEntry e = new TileEntry();
         initEntry(e, rs);
 
-        //load all the tile matrix entries
+        // load all the tile matrix entries (and join with the data table to see if a certain level
+        // has tiles available, given the indexes in the data table, it should be real quick)
         PreparedStatement psm = cx.prepareStatement(format(
-            "SELECT * FROM %s" + 
+            "SELECT *, exists(SELECT 1 FROM %s data where data.zoom_level = tileMatrix.zoom_level) as has_tiles" +
+            " FROM %s as tileMatrix" + 
             " WHERE table_name = ?" +
-            " ORDER BY zoom_level ASC", TILE_MATRIX_METADATA));
+            " ORDER BY zoom_level ASC", e.getTableName(), TILE_MATRIX_METADATA));
```

Comment:
does this need to be quoted, ah probably not because it is a prepared statement.